### ISSUE TITLE
Laravel 10 - Notifications not sending?

### ISSUE THREAD

**Issue by larson-carter** [laravel-notification-channels/apn#148]
Hello all,

I have my project on Laravel 10, I have a imessage clone app (sorta). I am unable to get the notifications to send to an actual device.

I have tried various things, followed the docs and tried to get ChatGPT to help.

I currently have my p8 key being used along with my User Model being the notifiable model. (Is this possibly the issue?)

Thanks for any insights. I'm obviously not being that detailed as I'm simply lost and I can admit that.

**Comment by dwightwatson** [comment]
Hey Larson. This channel is a slim wrapper around `edamov/pushok` so you best bet is to try and debug what is going on there. Laravel channels aren't very good at highlighting issues that occur under the hood - typically they dispatch a `NotificationFailed` event. We should probably add that to our own docs to help there.

In the meantime - I'd try dropping a `dd($responses)` inside [`ApnChannel`](https://github.com/laravel-notification-channels/apn/blob/master/src/ApnChannel.php#L38) and taking a look at the responses from `pushok` - or try `dd` in a few places of the channel to inspect what is going on.

Hope some of that helps - let me know if you are still running into trouble. Good luck!

**Comment by larson-carter** [comment]
@dwightwatson Thanks for this reply. 

I will try doing exactly as you stated to see if that helps. I have used ChatGPT to generate a test but my most recent problem when trying to run the test is: 
`InvalidArgumentException: Driver [apn] not supported. in file /Users/larsoncarter/Herd/CollegeNest/vendor/laravel/framework/src/Illuminate/Support/Manager.php on line 109`

**Comment by dwightwatson** [comment]
Check the docs out - you pass the class name of ApnChannel into the "via" method of your notification not the string "apn". Maybe it used to work that way and ChatGPT is remembering an older version of this package or it's just making it up. 😅

**Comment by larson-carter** [comment]
@dwightwatson What a good catch. ChatGPT had everything right so far except that. (I cross referenced my model and controller from the docs)

Now I'm getting: `Exception: unable to set private key file: &#039;/Users/larsoncarter/Herd/CollegeNest/storage/app/apn/aps.pem&#039; type PEM in file /Users/larsoncarter/Herd/CollegeNest/vendor/edamov/pushok/src/Client.php on line 161` which might just be a file conversion issue? Nonetheless I'll keep debugging unless you know the issue.

**Comment by dwightwatson** [comment]
Yeah, or sounds like it could be a permissions issue on the storage directory as well. You may need to reference the Laravel docs just to check what the directory should be set to. You can also try dropping "dd" inside the Client file referenced there to see if you can unearth where it's going wrong.

**Comment by larson-carter** [comment]
@dwightwatson UPDATE:

1.) I got everything integrated as it should be.
2.) Wrote a test to send notifications, it passes test however it does not actually send the notification to the device.

I have debugged and validated the .p8 cert, I have validated the device token, I have also logged to ensure it is sending to the proper user and the correct token, yet not avail.

**Comment by larson-carter** [comment]
@dwightwatson Sorry for being annoying but I just found this:

`The specified device token is invalid. Verify that the request contains a valid token and that the token matches the environment.`

That is what apple shows in their CloudKit Push notification panel when I try to manually to send a notification in the prod environment?

**Comment by dwightwatson** [comment]
Cool - double check how you're getting push tokens from your iOS device. Test out a raw token in the CloudKit debugger first to make sure it's all good before trying to plug it into Laravel.

**Comment by larson-carter** [comment]
I figured out the issue. It was due to the code not being able to read in the .ENV variables in config/broadcasting.php?

Seems like it only works if you enter in the vars directly into that file. Any other known fix to get it where I can pull it from the .ENV?

**Comment by dwightwatson** [comment]
Sounds like you're fighting the framework there. In Laravel the config files should be able to res from .ENV just using the env helper like shown in the docs. 

A common gotcha for newbies is running php artisan config:cache locally which caches all the config and won't read any changes. You can try php artisan config:clear to ensure there is no cache in place. Then run php artisan tinker and type in config('broadcasting') to see the output and what values Laravel is loading in.

**Comment by larson-carter** [comment]
@dwightwatson I spent some time to get it manually working with pusher. I then went back to this package here and dove into the vendor side of things and logged this response:

`[2024-09-02 08:46:18] local.INFO: APN Notification: Response Details {"status_code":403,"reason_phrase":"There was an error with the certificate or with the provider authentication token.","error_reason":"InvalidProviderToken","error_description":"The provider token is not valid or the token signature could not be verified","success":false}`

**Comment by dwightwatson** [comment]
Unfortunately there's not really enough there to go on - unless you're able to use that token and signature combination directly and see if it works without going through the channel. You could also test out using [`pushok`](https://github.com/edamov/pushok) standalone and see if you have any luck that way. Like I mentioned, this channel is really a thin wrapper around `pushok` to integrate it into Laravel's notification system.

**Comment by larson-carter** [comment]
@dwightwatson 

Thanks for the debugging tips. Ultimately I figured it out. Just had to move the broadcasting.php config and hard code in the environment stuff, even though via tinker it was properly populating.

I also had to tweak some of the feedback settings to ensure that everything was working properly. I guess that is one thing that I think would be neat for this repo is verbose responses and logging?

Anyways ultimately fixed and works flawlessly. Thank you very much!
